### DISCUSSION TITLE
1.3" OLED

### DISCUSSION THREAD

**Discussion by andrewerrington** [sethvoltz/OctoPrint-DisplayPanel#86] [Ideas]
I decided I'd like a slightly bigger display.

I purchased this:
https://m.aliexpress.com/item/4001205659261.html

It is a 1.3" OLED display with SSD1306 driver (same as the original 0.96" display), I2C interface, and 3.3V power supply.

It appears at address 0x3C, so far, so good, but only the top line of the display works, and it's slightly offset. The rest of the screen is just noise.

![IMG_20220125_164643~2](https://user-images.githubusercontent.com/3054284/150909904-a625225b-486d-4b4d-ad23-a14a7fd19892.jpg)

I'm happy to do some work on this, but if anyone has any obvious suggestions to start with I'd love to hear them.

**Comment by sethvoltz** [comment]
Hmm, my first guess was that the dimensions pixel dimensions were different as, the way I understand it, the graphics lib just spits out a long sequence of bytes that auto-wrap to create the image. However, the listing claims the same dimensions as expected. 

That said, even the first line isn't working there and the text that is working is way off.

**Comment by kgutwin** [comment]
I'm wondering whether or not you can get it to work with some of the other examples out there?

I found this page which walks you through running some examples from the Adafruit library. Could you test one of these to see if they work?

https://www.raspberrypi-spy.co.uk/2018/04/i2c-oled-display-module-with-raspberry-pi/

**Comment by andrewerrington** [comment]
Yup. I'm happy to try the OLED test code. Naturally, being lazy, I was hoping for a drop-in replacement. It seems that some of the other 1.3" displays use a different driver chip. It's also possible that the vendor's description was, shall we say, inaccurate.

**Reply by sethvoltz** [reply]
It's happened before. I would also double check all your wiring. We've seen issues where poorly crimped wired can cause issues with these displays. For my final setup, I ended up crimping all my own ends and soldering a custom socket for the Pi side.

**Reply by andrewerrington** [reply]
It's not the wiring. I have a 0.96" module which I swapped in and it works (of course).

I've tried a few debugging things, but I think the problem is in the Adafruit driver. I also think the module might have an SH1106 driver, despite the vendor's claim.

I'll try some tests with another SSD1306 driver (not Adafruit) and an SH1106 driver too.

**Comment by andrewerrington** [comment]
Ok. A little progress, based on random Googling.

In the Adafruit driver there is a class _SSD1306, which has a parameter page_addressing. I forced this to True, because reasons, and I have an interim result.

![IMG_20220126_131400~2](https://user-images.githubusercontent.com/3054284/151082507-61f4d9ad-39a7-479b-9e61-2d0efe352bd7.jpg)

You can see four pixels of noise in the centre of the display, which is dead memory. The SH1106 chip supports up to 132 pixels wide, but the display itself is 128 pixels. The SSD1306 chip only supports 128 pixels. The generated image is 128 pixels wide, so the last four pixels are never written. Obviously the display should be shifted so that the dead pixels are not visible.

Word on the street is that the SH1106 based displays which have 128 pixels are offset to the "centre". i.e. there is a strip of two pixels to the left and right which we don't see. Other drivers deal with this, but the Adafruit one might not.

**Reply by sethvoltz** [reply]
I've been considering a switch to a different driver for a number of reasons, but I never went back. 

Now that the code has been refactored with the virtual and physical displays being separate, maybe now is the time.

**Comment by andrewerrington** [comment]
Don't get too excited. It's a cheap hack of which I am not proud, but it shows that the hardware can be made to work:
![IMG_20220126_211604~2](https://user-images.githubusercontent.com/3054284/151127663-6f806861-a62d-4239-b368-4fd9bd013d52.jpg)

I don't have any buttons yet, but the OLED display changes mode when I click on the buttons on the web version of the display.

I've posted on the Adafruit webpage for the driver asking about some hints. Hopefully I can make it work in an elegant way.

If you wanted to investigate new drivers I'm not sure what to recommend. I think u8g2 might be a good candidate, but I'm not sure it has been ported to Python.

While you are thinking of that you might consider deprecating RPi.GPIO and moving to gpiozero. I think that is the current best-supported GPIO library.

**Reply by andrewerrington** [reply]
Ok, there's this

https://github.com/ribasco/u8g2-rpi-demo

But it uses Wiring Pi, which is definitely deprecated, and only supports SPI.

Other information here:
https://github.com/olikraus/u8g2/wiki/Porting-to-new-MCU-platform

**Reply by andrewerrington** [reply]
There is also this, which is ported to Python:
https://github.com/rm-hull/luma.oled

**Reply by sethvoltz** [reply]
What was the ugly hack to make it work? Do you know if it works with the original 0.9" OLEDs as well?

Either way, if it's not _too_ ugly a hack, would you mind putting together a PR with the changes?

**Reply by andrewerrington** [reply]
Ha ha. No.

First, I set the width to 132 pixels, so that I would get a bigger buffer, and turned page addressing mode on. Next I modified the Adafruit code to send the correct startup code if the display was 132x64 instead of 128x64. Finally, when each 'strip' of bytes was sent to the display I chopped the first half and last half and the extra four bytes and stuck the pieces back in a new order so they looked right on the screen. Except it's not quite right as some text is now centred in a 132-pixel width instead of a 128-pixel width.

To call it "unsatisfactory" would be an understatement. And it won't work on the original display because the output has been chopped up and rearranged (although I haven't confirmed that).

The luma project I linked to is supposed to support SH1106 directly, so I'll try that out. If I can seamlessly switch between the two different models by changing the display declaration I'll be happy.

**Reply by andrewerrington** [reply]
Ok, following the instructions [here](https://community.octoprint.org/t/octopi-how-to-use-pip-through-ssh/24111) I was able to use pip to install luma into Octoprint's environment. A couple of minutes later I had modified the startup code for micro_panel.py to display "Hello World" with basically zero effort. So the SH1106 display works, and the library works. I think the programming interface is different to the Adafruit interface, so I don't know how much work is needed to change that around. Hopefully I just need to change some low-level routines and everything on top will Just Work.

My next experiment is to install the older SSD1306 display, change one line of code, and see that the "Hello World" message appears. Long term I think that would be useful to others to support both of these displays, but luma also supports several other models.

luma also has a virtual display which mimics the hardware. If you decide to use luma maybe you can take advantage of that in virtual_panel.py.

**Reply by sethvoltz** [reply]
Yeah, if Luma works with different chipsets and we can get the same graphics working on it, that would be a nice extension to be able to select the chipset from a dropdown in the prefs, defaulting to the original display so as to not mess up existing installations.

**Reply by andrewerrington** [reply]
Well, it's up to you, but you probably have the whole architecture in your head so you might be able to see how easily this might fit in (or not). I'll carry on with my experiments later and report back.

**Reply by andrewerrington** [reply]
Ok. SSD1306 display with SSD1306 driver:
![IMG_20220127_125654~2](https://user-images.githubusercontent.com/3054284/151267290-f8380fb1-bc3e-4023-b997-784b01c9503d.jpg)

SSD1306 display with SH1106 driver:
![IMG_20220127_125834~2](https://user-images.githubusercontent.com/3054284/151267346-55cbdbd1-420a-48e8-8221-04bcd2ea94e9.jpg)

SH1106 display with SSD1306 driver:
![IMG_20220127_130045~2](https://user-images.githubusercontent.com/3054284/151267401-698e3165-10bd-410c-940a-14445e0d1553.jpg)

SH1106 display with SH1106 driver:
![IMG_20220127_125942~2](https://user-images.githubusercontent.com/3054284/151267471-89f52f09-872f-4776-a3e1-efe2e42d0572.jpg)

It's only a simple demo, but I only have to change one line to declare the type of display in use. You can see that the displays don't work quite right with the wrong driver, despite there being some commonality.

**Reply by andrewerrington** [reply]
Ta da!
![IMG_20220127_134256~2](https://user-images.githubusercontent.com/3054284/151270772-6412b740-31b1-4889-ad3b-c69059592420.jpg)

I suppose I'd better figure out how I did it, and document it. Rest assured, I did the absolute minimum necessary to make it work.

That font, though.

**Reply by kgutwin** [reply]
@andrewerrington Here's some hints on the architecture that might help you implement a new panel driver:

* The code is designed to support multiple different types of "panels" (where a panel is defined as both the display and the buttons). You can create a new file under `octoprint_display_panel/panels/` and declare a class that has the methods `shutdown()`, `fill()`, `image()`, `show()`, `poweroff()`, and `poweron()`. The `setup()` method is optional but you'll probably want that.
* Since you are probably using the same sort of input method as the standard panel (GPIOs for buttons), you may want to just subclass the existing `MicroPanel` class from `micro_panel.py`. Unfortunately, the GPIO initialization is currently mixed in with the display initialization, so it would make sense to first do a little refactoring there to split those two apart.
* The main way it works is the `Panels` class in `octoprint_display_panel/panels/__init__.py` is a proxy for any and all panels connected to the device. All panels will have the same size and show the same content. The `screens` portion of the code generates the screen images, and feeds it into the `image()` method, which gets passed down to the actual classes that implement the panels. Your code simply needs to be able to take the provided bitmap image and push it to the screen.
* To add a new panel type, you'll want to define a new setting (look through `octoprint_display_panel/__init__.py` and other spots for how to do that) and then in the `__init__()` method of the `Panels` class, check that setting to determine which panel type to initialize. You should gracefully handle the case when your panel can't initialize, for example if the `board` module fails to load. (This was important for me while working on the refactor, as I was developing using my laptop rather than directly on the Pi.)

Feel free to ask any other questions you may have!

**Reply by sethvoltz** [reply]
@andrewerrington huzzah! Good to see the driver is doing the right thing. 

> That font, though.

Haha, yeah, it's pretty terrible but it works and it doesn't require any additional font loading. Definitely take a look at what @kgutwin said and feel free to ask questions. We can take a look at PRs and it would be pretty great to have the selector to choose display types.

**Reply by andrewerrington** [reply]
I'll just paste it here. Basically, follow the instructions using pip to install luma.oled, then use this version of micro_panel.py

If you diff with the original you will see that I just removed adafruit's library and create a very similar display object with luma.oled

I have tested self.disp = sh1106(self.serial) and self.disp = ssd1306(self.serial)

I have modified some of the object methods because they were tailored to the Adafruit library, and the luma.oled library does things differently. Although, fundamentally, it's just "create an image with pillow" then "send image to hardware". I have not tested all of the methods. I just modified them to sensible commands that ought to work.

I also put a tiny test routine at the beginning, after the OLED object is created, but before it is used, with the idea that if it works then it indicates a rudimentary level of functionality before any other code runs. It draws a box around the very edge of the display, with some text in the centre.

I'm happy that I can use the larger display now, but there are a bunch of 128x64 displays with larger physical sizes that could now be used if they needed a different driver to SSH1306.

I've also ordered some of these:
https://www.aliexpress.com/item/1005003560941062.html
It's a button board about the same width as the 1.3" OLED PCB.

I was getting fed up of opening a web browser and logging in just to kill a job, or get Octoprint to establish a serial connection, so I wanted something exactly like this project. Thanks.

`
from luma.core.interface.serial import i2c
from luma.core.render import canvas
from luma.oled.device import ssd1306, sh1106
import RPi.GPIO as GPIO
import time

def bcm2board(bcm_pin):
    pinmap = [-1, -1, -1,  7, 29, 31, -1, -1, -1, -1, -1, 32,
              33, -1, -1, 36, 11, 12, 35, 38, 40, 15, 16, 18,
              22, 37, 13]
    if bcm_pin != -1:
        return pinmap[bcm_pin - 1]
    return -1

class MicroPanel:
    """Interface to the standard I2C and GPIO-driven Micro Panel.
    """
    width = 128
    height = 64

    def __init__(self, button_callback):
        self.button_event_callback = button_callback
        self.gpio_pinset = set()

    def setup(self, settings):
        """Apply settings from OctoPrint's SettingsPlugin mixin to
        configure the panel.
        """
        self.i2c_address = int(settings.get(['i2c_address'], merged=True), 0)
        self.input_pinset = {
            settings.get_int([f'pin_{p}'], merged=True): p
            for p in ['cancel', 'mode', 'pause', 'play']
        }
        self.debounce_time = settings.get_int(['debounce'], merged=True)

        # set up display
        self.serial = i2c(port=1, address=self.i2c_address)
        self.disp = sh1106(self.serial)
        #self.disp = ssd1306(self.serial)

        # set up GPIO mode
        current_mode = GPIO.getmode()
        if current_mode is None:
            # set GPIO to BCM numbering
            GPIO.setmode(GPIO.BCM)

        elif current_mode != GPIO.BCM:
            # remap to BOARD numbering
            GPIO.setmode(current_mode)
            self.input_pinset = {bcm2board(p): l
                                 for p, l in self.input_pinset.items()}
        GPIO.setwarnings(False)

        # set up pins
        for gpio_pin in self.input_pinset:
            if gpio_pin == -1:
                continue

            GPIO.setup(gpio_pin, GPIO.IN, GPIO.PUD_UP)
            GPIO.remove_event_detect(gpio_pin)
            GPIO.add_event_detect(gpio_pin, GPIO.FALLING,
                                  callback=self.handle_gpio_event,
                                  bouncetime=self.debounce_time)
            self.gpio_pinset.add(gpio_pin)

        # clean up any pins that may not be selected any more
        cleaned_pins = set()
        for gpio_pin in self.gpio_pinset.difference(self.input_pinset.keys()):
            try:
                GPIO.remove_event_detect(gpio_pin)
                GPIO.cleanup(gpio_pin)
            except:
                logger.exception(f'failed to clean up GPIO pin {gpio_pin}')
            else:
                cleaned_pins.add(gpio_pin)
        self.gpio_pinset.difference_update(cleaned_pins)

        # Startup OLED test
        with canvas(self.disp) as draw:
            draw.rectangle(self.disp.bounding_box, outline="white", fill="black")
            draw.text((38,20),"OctoPrint", fill="white")
            draw.text((31,36),"Micro Panel", fill="white")

        time.sleep(2)
        self.disp.clear()

    def shutdown(self):
        """Called during plugin shutdown.
        """
        for gpio_pin in self.input_pinset:
            if gpio_pin == -1:
                continue
            GPIO.remove_event_detect(gpio_pin)
            GPIO.cleanup(gpio_pin)

    def fill(self, v):
        """Fill the screen with the specified color.
        """
        with canvas(self.disp) as draw:
            draw.rectangle(self.disp.bounding_box, outline=v, fill=v)

    def image(self, img):
        """Set an image to be shown on screen.
        """
        self.disp.display(img)

    def show(self):
        """Show the currently set image on the screen.
        """
        self.disp.show()

    def poweroff(self):
        """Turn the display off.
        """
        self.disp.hide()

    def poweron(self):
        """Turn the display on.
        """
        self.disp.show()

    def handle_gpio_event(self, channel):
        """Called on a GPIO event, translate an input channel to a button label
        and invokes the button callback function with that label.
        """
        if channel not in self.input_pinset:
            return

        self.button_event_callback(self.input_pinset[channel])
`

**Reply by andrewerrington** [reply]
Hmm. Github doesn't like my code formatting. Oh well.

**Reply by andrewerrington** [reply]
Also, the show() method should be removed, otherwise screen blanking doesn't work.

`
def show(self):
    """Show the currently set image on the screen.
    """
    pass
`

It's a bit tightly integrated in other code, which I didn't want to mess with.

And GitHub messes up the formatting again.

**Reply by slipslyman** [reply]
@andrewerrington thanks for the ping on my previous post, hoping you might be able to help here. 

i2c detect -y 1 returns an address, either a 3C or 3D, wiring as per Step 3 and the comments section respectively of this  https://www.instructables.com/OLED-Tutorial-Convert-SPI-to-I2C/ 

My particular sh1106 driver oled display  is the 6 pin one below so I've done a bit of troubleshooting to check I've not got a simple wiring issue. 

![274576873_291400429763781_2464343627469295243_n](https://user-images.githubusercontent.com/93126736/156836331-4c54ed79-8573-4420-9a19-631f5a0018e2.jpg)

I've installed luma.oled as per these instructions, https://luma-oled.readthedocs.io/en/latest/software.html. I can get myself a lovely "hello world" displayed following the example a bit further on in that link. 

![274635094_356032643062858_6210021358264267178_n](https://user-images.githubusercontent.com/93126736/156836339-ff4a3a12-466d-4190-9b7c-b39276c3cf13.jpg)

Thought I was  on the home straight at this point, I've modified the micro_panel.py to replace the adafruit library with the luma.oled and removed the show() method as you've outlined above, but now I only get a blank screen. Not even the joy of the garbled static anymore. 

I've diff'd your micro_panel.py and the original.  The only sections  I can see changed to accommodate luma are: 

from luma.core.interface.serial import i2c
from luma.core.render import canvas
from luma.oled.device import ssd1306, sh1106
import RPi.GPIO as GPIO
import time

and 

        # set up display
        self.serial = i2c(port=1, address=self.i2c_address)
        self.disp = sh1106(self.serial)
        #self.disp = ssd1306(self.serial)

Additionally I've looked through the __init__.py to see if I missed some other change referencing luma there but no dice. 

Admittedly I'm super green to all of this, so for full transparency of where I might be arsing it up, I'm using nano to edit the micro_panel.py hoping that's the correct way. My last suspicion is I haven't installed luma.oled correctly in the oprint environment

**Reply by andrewerrington** [reply]
Hi there!

The only thing that comes to mind is did you install Luma onto the Pi in the OS, or onto the Pi into Octoprint's environment? Octoprint maintains its own drivers and files independently of the OS-installed ones.

The instructions here tell you how to use pip to install into Octoprint's environment are here:
https://community.octoprint.org/t/octopi-how-to-use-pip-through-ssh/24111

Otherwise, sorry. I suppose one day it might all get packaged up nicely for easy installation.

**Reply by slipslyman** [reply]
I did end up trying to make sure I had Luma installed in the correct environment. 

No success unfortunately. I suspect this is the likely reason though.

Tested my skills and paitence a bit too much and opted for a 3.5" tft display which I miraculously got to work. 

Thanks though, definitely pushed me to learn a bit more so still came away with something.

**Reply by andrewerrington** [reply]
Ok. Cool. If you don't mind, could you start a new discussion with some key points about how you did it (or put them here) so the next person has a head start? No worries if you choose not to (especially as you might be expected to support it) but useful if you do.

**Comment by andrewerrington** [comment]
I thought I had written a conclusion here, but it does not seem to be present.

I built a four-button board the same size as the ones I have ordered. Everything works as expected.

![IMG_20220128_160936~2](https://user-images.githubusercontent.com/3054284/151630689-b222e6c5-0bf1-4043-b81f-2d7e02e08dd7.jpg)

There are occasional glitches when updating the display, but that could be because the core code is written to use the Adafruit library which may deal with screen updates in a different way.

Going forward I'm not sure what the best approach is. I think the luma.oled library is more versatile than the Adafruit library. Upgrading from one to the other might be tricky though, so maybe a new project could be started, with pretty much the same code, but installs the new library.

Anyway, I'm going to build an enclosure for my screen and buttons and install one each on my two printers. Great project!

**Reply by andrewerrington** [reply]
And for my own memory I'm planning to use Cat5 Ethernet cable to connect the display board with the Pi. Two reasons for this. Firstly because I don't know where I'm going to put the display and the Pi. And secondly, I couldn't help but notice there are 8 connections.

1 Button 1
2 Button 2
3 +3.3V
4 SCL
5 GND
6 SDA
7 Button 3
8 Button 4

On the display side GND will connect to the OLED display and the common pin for the buttons. As you can see, SCL and GND are on a pair together, and SDA and power are on another pair together, although for the short distances I am planning (about 0.5m) it's probably not an issue.

**Reply by kgutwin** [reply]
Some quick recommendations from my experience - despite having the panel and Pi close together (about 6 inches) I had a lot of frustrating interference issues, with display glitches and phantom button presses. It seemed particularly tied to certain stepper motors for me (Ender 3). What I did to solve it:

* Split the display and button grounds into two separate wires going to two different GND pins on the Pi. 
* Added outboard pull-up resistors on the button lines, as close as possible to the Pi, to make up for the relatively weak internal pull-ups on the Pi’s GPIOs. 

I was really worried I was going to need to add shielding or go to more drastic measures to cut down on EMI from the printer, but those two steps seemed to solve it for me. I didn’t try them separately so I can’t say if one was more impactful than the other, unfortunately…

**Reply by andrewerrington** [reply]
Thanks for that. Things are a little worse with 0.5m Cat5, but it still works. I could try a decoupling capacitor on the display too.

The problem is that I have two different printers, which means two different mounting solutions for the Pi itself and the display. Ideally I'd like the display to be in (or part of) the Pi case, but I haven't looked to see if a single design would work for both printers. Currently both Pis are on the desk tucked away behind the printers.

I've found some mounting designs on Thingiverse but I need to evaluate them. One printer is an Ender 3 V2, the other is a Monoprice Mini.

Sigh. First-world problems.

**Reply by sethvoltz** [reply]
I was worried about these things as well but I have about a meter of Cat5E and rarely, if ever, see glitching on the display. I2C is a pretty resilient wire protocol. The buttons have had issues, though. Never with false presses but instead not registering clicks. I have to hold them a tick longer than we normally would for a guaranteed hit. 

The things I did do were solder all the connections for a rock solid interface and also crimp the Pi ends into a custom connector. I don't have my notes handy but if you'd like more info I can dig it out. 

Things I could do to improve this would be decoupling caps and some external pull-ups on the buttons.

**Comment by andrewerrington** [comment]
I see there is one other person looking for this. What is the best way to proceed?
